flink-runtime/src/test/java/org/apache/flink/runtime/testutils/InMemoryReporter.java

```diff
@@ -50,6 +50,11 @@
  *
  * <p>This class can only be accessed through {@link InMemoryReporterRule#getReporter()} to ensure
  * that test cases are properly isolated.
+ *
+ * <p>Note that at this time, there is not a strong guarantee that metrics from one job in test case
+ * A cannot spill over to a job from test case B if both test cases use the same minicluster - even
+ * when run sequentially. To ensure that assertions against metrics are stable, use rather unique
+ * task and operator names and respective metric patterns.
```

Comment:
I have to admit that the `ThreadLocal` variable confused me a bit. I think it only works because the `MiniCluster.start` is called from the test thread. But if this is not the case for whatever reason, then all the tests will break.

Reply:
I created FLINK-23803 as a follow-up on this one as it feels to be out of scope of the issue at hand.